mkosi/util.py

```diff
@@ -319,3 +320,15 @@ def try_or(fn: Callable[..., T], exception: type[Exception], default: T) -> T:
         return fn()
     except exception:
         return default
+
+
+def groupby(seq: Sequence[T], key: Callable[[T], S]) -> Sequence[tuple[S, list[T]]]:
```

Comment:
We can't make the promise for the return value, so 
```suggestion
def groupby(seq: Sequence[T], key: Callable[[T], S]) -> list[tuple[S, list[T]]]:
```
would be a better return.

The name `groupby` is pretty fitting, but I think something like `group_by_function` would have lesser change to be mistaken by cursory reading with `itertools.groupby`.

Reply:
To be fair, `itertools.groupby` also takes a `key` function and this could be written using `itertools.groupby`. Giving this a better name is hard.

Reply:
We should be fine with groupby(). I don't wanna bother with itertools.groupby() because then we have to sort and it doesn't end up more readable then the version without itertools.groupby().